### PR TITLE
Support a generalized way to configure Liquibase global properties through Spring Boot configuration

### LINKED ISSUE #47212: Proposal: Support a generalized way to configure Liquibase global properties through Spring Boot configuration
**Problem**  
Liquibase has an expanding set of [global configuration options](https://docs.liquibase.com/parameters/home.html) (e.g. `duplicateFileMode`, `logLevel`). Currently, these can only be set via environment variables or JVM system properties (e.g. `-Dliquibase.duplicateFileMode=WARN`). 

In a Spring Boot application, this creates inconsistency. Most Liquibase settings live under `spring.liquibase.*`(`application.yml`) but in order to set global properties we must set global values elsewhere, such as through system properties, env vars, or through `Customizer<Liquibase>`. In my own experience, this fragmented approach made it difficult to configure liquibase across environments and made testing more difficult. 

**Use case**  
My personal use case was that we had multiple Liquibase changelog files and frequently saw an error indicating that Liquibase found duplicates. The recommended fix is to set `liquibase.duplicateFileMode=WARN`. The workaround we used was to set the variable in our Gradle file, which worked but led to issues I mentioned above.

**Proposal**  
Introduce a **generic pass-through mechanism** for Liquibase global properties. This was the proposed approach on the long-standing Liquibase issue [here](https://github.com/liquibase/liquibase/issues/3230#issuecomment-1329824052). For example:

```yaml
spring:
  liquibase:
    change-log: classpath:/db/changelog_master.xml
    globals:
      duplicate-file-mode: WARN
      log-level: INFO
```
**Behavior:**

- At startup, entries under `spring.liquibase.globals.*` are applied to Liquibase’s global configuration before SpringLiquibase executes (internally via the existing customizer / LiquibaseConfiguration bridge).
- Spring Boot does not hard-code or validate the option names; it simply passes them through. Liquibase remains the source of truth for supported keys and semantics.

**Benefits**

- Keeps all Liquibase config in `application.yml` as the single source of truth
- Reduces production/test mismatches.
- Minimal maintenance burden: Spring Boot simply passes properties through; Liquibase defines the option set.

**Proof of concept**

I have a Kotlin project where I constructed a solution for this. It works, though there might be a better place to hook in to access the spring configuration. Note that I am not using `spring.liquibase.globals` because `spring.liquibase` maps to a static properties class. To implement that behavior, you'd need to add globals as a string map property. If this looks like something that your team would be interested in, I'd be happy to do the implementation or adapt it to be more consistent with the integration code.

src/main/resources/META-INF/services/liquibase.configuration.ConfigurationValueProvider
```
com.example.liquibase.config.SpringLiquibaseGlobalConfigProvider
```

src/main/resources/META-INF/spring.factories
```
org.springframework.context.ApplicationContextInitializer=\
com.example.liquibase.config.LiquibaseEnvironmentInitializer
```

src/main/kotlin/com/example/liquibase/config/LiquibaseEnvironmentInitializer.kt
```kotlin
package com.example.liquibase.config

import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.core.Ordered

class LiquibaseEnvironmentInitializer :
    ApplicationContextInitializer<ConfigurableApplicationContext>, Ordered {

    override fun initialize(applicationContext: ConfigurableApplicationContext) {
        SpringLiquibaseGlobalConfigProvider.environment = applicationContext.environment
    }

    override fun getOrder(): Int = Ordered.HIGHEST_PRECEDENCE
}
```

src/main/kotlin/com/example/liquibase/config/SpringLiquibaseGlobalConfigProvider.kt
```kotlin
package com.example.liquibase.config

import liquibase.configuration.AbstractConfigurationValueProvider
import liquibase.configuration.ProvidedValue
import org.springframework.core.env.ConfigurableEnvironment

class SpringLiquibaseGlobalConfigProvider : AbstractConfigurationValueProvider() {

    companion object {
        var environment: ConfigurableEnvironment? = null
    }

    override fun getPrecedence() = 10_000

    override fun getProvidedValue(vararg keyAndAliases: String?): ProvidedValue? {
        val env = environment ?: return null
        for (key in keyAndAliases.filterNotNull()) {
            val shortKey = key.substringAfterLast(".")
            val candidates = listOf(
                "app.liquibase.global-config.$shortKey",
                "app.liquibase.global-config.${toKebabCase(shortKey)}"
            )
            for (candidate in candidates) {
                env.getProperty(candidate)?.let { value ->
                    return ProvidedValue(key, key, value, "Spring Environment ($candidate)", this)
                }
            }
        }
        return null
    }

    private fun toKebabCase(s: String): String =
        s.foldIndexed(StringBuilder()) { i, acc, ch ->
            if (ch.isUpperCase() && i > 0) acc.append('-')
            acc.append(ch.lowercaseChar())
        }.toString()
}
```

### PR DESCRIPTION
**Pull request by dmiska25**
## Summary
This PR introduces support for configuring Liquibase’s **global properties** directly through Spring Boot’s configuration system.  

A new map property has been added:  

```yaml
spring:
  liquibase:
    properties:
      liquibase.duplicateFileMode: 'WARN'
```

At startup, these entries are passed through to Liquibase’s `ConfigurationValueProvider`. This aligns with how Spring Boot already supports pass-through properties for JPA (`spring.jpa.properties.*`).

---

## Motivation
Liquibase supports an expanding set of global configuration options (e.g., `duplicateFileMode`, `logLevel`, `searchPath`). Until now, these could only be set using JVM system properties or environment variables if they were not specified in the Spring Boot configuration.  

That approach created fragmentation: most Liquibase settings were configured under `spring.liquibase.*`, while global properties had to be managed separately. This change unifies configuration, allowing Spring users to manage all Liquibase options from one place.  

My motivating use case: we had multiple Liquibase changelog files and frequently saw an error indicating duplicates. The recommended fix was to set `liquibase.duplicateFileMode=WARN`. The workaround was to set it in Gradle system properties, but that made testing and environment alignment harder.

---

## Design
- A new `EnvironmentConfigurationValueProvider` bridges Spring’s `Environment` to Liquibase’s configuration system.  
- Properties under `spring.liquibase.properties.*` are read exactly as-is (no relaxed binding or rewriting).  
- Liquibase’s own precedence rules still apply: system properties and environment variables can override these values.  
- The provider is registered programmatically via Boot’s auto-configuration lifecycle, ensuring it is active before Liquibase runs.  

### Precedence Choice
The provider is registered with a precedence of **100**, above defaults but below environment variables.

| Provider                                   | Precedence |
|--------------------------------------------|------------|
| DefaultsFileValueProvider                   | 50         |
| DefaultsFileValueProvider (local)           | 51         |
| **EnvironmentConfigurationValueProvider** (new) | **100**     |
| Environment Variables (LIQUIBASE_*)        | 150        |
| System Properties                          | ~200 (varies by version) |
| Command-line arguments (`--`)              | 250        |
| DeprecatedConfigurationValueProvider       | 350        |
| InteractivePromptingValueProvider          | 500        |

Reference: [Liquibase parameters documentation](https://docs.liquibase.com/reference-guide/parameters/what-are-parameters).  

---

## Testing
- **Unit tests** validate provider behavior (lookup, precedence, null handling).  
- **Integration tests** ensure properties are applied during context bootstrap.  
- Verified manually by building a new sample project and running Liquibase migrations with global properties set via `application.yml` and `application.properties`.  

---

## Documentation
- Updated Javadoc for `spring.liquibase.properties` to clarify usage, with an example key.  
- Liquibase’s own [documentation](https://contribute.liquibase.com/extensions-integrations/directory/integration-docs/springboot/configuration/) covers the supported keys and semantics. I also plan to open a PR on their side to reference this integration.

---

## Implementation Notes
1. To avoid cross-test contamination, existing provider instances are unregistered before a new one is registered.  
2. Chose Javadoc + appendix documentation for succinctness. Open to expanding on the liquibase how-to if maintainers prefer or update guides.  

Fixes #47212

### CODE REVIEW

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java [hunk 1]
```diff
@@ -188,6 +190,24 @@ private void applyConnectionDetails(LiquibaseConnectionDetails connectionDetails
 
 	}
 
+	@Bean
+	static BeanFactoryPostProcessor liquibaseConfigurationValueProviderRegistrar(Environment environment) {
```
Comment (wilkinsona) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: Rather than using a `BFPP`, I think this could be done in the `@Bean` method that creates the `SpringLiquibase` instance. This will ensure it only applies when auto-configuring Liquibase and will back off when the user wants to exert more control.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java [hunk 2]
```diff
@@ -188,6 +190,24 @@ private void applyConnectionDetails(LiquibaseConnectionDetails connectionDetails
...
+
+		return (beanFactory) -> {
+			var liquibaseConfiguration = liquibase.Scope.getCurrentScope()
```
Comment (wilkinsona) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: We don't use `var` in the Spring Boot codebase.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java [hunk 3]
```diff
@@ -188,6 +190,24 @@ private void applyConnectionDetails(LiquibaseConnectionDetails connectionDetails
...
+				.getSingleton(liquibase.configuration.LiquibaseConfiguration.class);
+
+			// Remove any previously registered instance of our provider class
+			liquibaseConfiguration.getProviders()
+				.stream()
+				.filter((provider) -> provider.getClass() == EnvironmentConfigurationValueProvider.class)
+				.toList()
+				.forEach(liquibaseConfiguration::unregisterProvider);
```
Comment (wilkinsona) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: Hopefully this won't be necessary when using a `@Bean` method. If it is necessary, one scenario where the `@Bean` method may be called multiple times is in tests with slightly different configuration that result in the creation of multiple contexts. Removing one provider in favour of a new one may cause problems there.

File: module/spring-boot-liquibase/src/test/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProviderTests.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+package org.springframework.boot.liquibase.autoconfigure;
+
+import liquibase.configuration.ProvidedValue;
+import org.junit.jupiter.api.Test;
+import org.springframework.mock.env.MockEnvironment;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class EnvironmentConfigurationValueProviderTests {
```
Comment (wilkinsona) [inline module/spring-boot-liquibase/src/test/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProviderTests.java:36]: Please add some class javadoc. See other `…Tests` classes for examples.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		for (String requestedKey : keyAndAliases) {
+			if (requestedKey == null) {
+				continue;
+			}
+			// Exact pass-through: no rewriting of the Liquibase key
+			String propertyName = PREFIX + requestedKey;
+			String value = this.environment.getProperty(propertyName);
+			if (value != null) {
+				// requestedKey: the key Liquibase asked for
+				// propertyName: the exact Spring Environment property we read
```
Comment (wilkinsona) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java]: I don't think these comments are necessary.

File: module/spring-boot-liquibase/src/test/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProviderTests.java (hunk 4 above)
Comment (djmiska25) [reply module/spring-boot-liquibase/src/test/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProviderTests.java:36]: Added. I wasn't sure if `@Author` is appropriate for community members, so I left that out; If you want me to set that for myself or you, I can do that.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java (hunk 3 above)
Comment (djmiska25) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: I added this mostly for tests. This is needed because the Liquibase configuration providers are static and when the LiquibaseAutoConfigurationTests run, they destroy Spring context in between tests, leaving the register with a stale configuration variable. I could handle this inside each test, but I think it might be wise to keep in source for downstream consumers too. I don't _think_ this would be an issue in production runs, but I'll lean on your experience for that.

File: module/spring-boot-liquibase/src/test/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProviderTests.java (hunk 4 above)
Comment (wilkinsona) [reply module/spring-boot-liquibase/src/test/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProviderTests.java:36]: Please do add an `@author` tag. It's absolutely appropriate for community members. It's warranted on the other classes that you have created or modified as well.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java (hunk 3 above)
Comment (wilkinsona) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: I wonder if we could avoid the problem by sub-classing `SpringLiquibase` and overriding `afterPropertiesSet` to create a child scope before calling `super`. This child scope could hold the `EnvironmentConfigurationValueProvider` and, hopefully, remove any problems of multiple contexts interfering with each other. Our existing subclass (`DataSourceClosingSpringLiquibase`) would become a sub-class of the proposed new sub-class.

Comment (djmiska25) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: So if I understand correctly, your suggestion is to execute the Spring Liquibase logic within a Liquibase child scope with the new provider defined? If so, I created this subclass like [this](https://gist.github.com/dmiska25/115bc758e53e4bdce4709a9c75284e40). But, in testing, what I found out is that Liquibase still registers its providers globally, even if the provider was registered to a child scope.

Comment (djmiska25) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: Ok, so I think one way it could be done is if the provider was defined by the scope(scope id passed in during construction). The provider can then request the current scope id at any point and then just return false for a lookup if its not the appropriate scope it was created with. My only thought is whether it's necessary to have this level of separation? Is it realistic that two Spring Liquibases would ever run at the same time? If it is, I could implement this, but if it's not, I'd lean to simplicity.

Comment (wilkinsona) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: Having two run in parallel is unlikely, but having them run serially with potentially different configuration is definitely possible both in tests and in apps with a context hierarchy. There's also the Actuator endpoint to consider where we use StandardChangeLogHistoryService at that can be called at any time, not just during application context refresh. I haven't examined the code path of the history service to know if it references configuration values that would be affected by configuration value providers.

Comment (djmiska25) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/LiquibaseAutoConfiguration.java]: Sorry for the delay, finally circling back to this. I updated as you suggested. I verified that there are a _few_ Liquibase properties(I added a test for one) that do get referenced in the endpoint, so I wrapped that as well. I verified that at least all the tests in the module are referencing the correct env through logging. It is possible for users to still do things like execute Liquibase themselves, but that's the only edge case I can think of, otherwise, that would not be wrapped currently. I added a few other comments for some other thoughts I had, like logging and fallbacks as well, that I'd love your input on.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		return 100;
+	}
+
+	@Override
+	public @Nullable ProvidedValue getProvidedValue(String... keyAndAliases) {
+		if (keyAndAliases == null) {
+			return null;
+		}
+
+		String environmentId = liquibase.Scope.getCurrentScope().get(SPRING_ENV_ID_KEY, String.class);
```
Comment (djmiska25) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java:73]: We could add additional strategies here, such as inferring the env if there's only one, for example. I opted for the simplest solution for now, but would love input on this.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+		if (environmentId == null) {
+			return null;
```
Comment (djmiska25) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java:75]: I considered adding logging here if we get a request for a config from this provider but we can't infer the env but I didn't see any other logging in the module, so I skipped for now.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java (hunk 6 above)
Comment (snicoll) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java:73]: I am not aware of how scope works but can't we just put the `Environment` and not have a local cache here?

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentAwareSpringLiquibase.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ * A {@link SpringLiquibase} subclass that creates a Liquibase child scope with a
+ * reference to the Spring Environment, allowing the
+ * {@link EnvironmentConfigurationValueProvider} to access the correct Environment for
+ * configuration values.
+ *
+ * @author Dylan Miska
+ */
+class EnvironmentAwareSpringLiquibase extends SpringLiquibase implements ApplicationContextAware, DisposableBean {
+
+	private @Nullable String environmentId;
```
Comment (snicoll) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentAwareSpringLiquibase.java:41]: It's not supposed to be nullable. I know that `ApplicationContext#getId` can return null but that's an oversight, see https://github.com/spring-projects/spring-framework/issues/35925

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentAwareSpringLiquibase.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	@Override
+	public void afterPropertiesSet() throws LiquibaseException {
+		try {
+			Map<String, Object> scopeValues = new HashMap<>();
+			scopeValues.put(EnvironmentConfigurationValueProvider.SPRING_ENV_ID_KEY, this.environmentId);
+			Scope.child(scopeValues, EnvironmentAwareSpringLiquibase.super::afterPropertiesSet);
+		}
+		catch (Exception ex) {
```
Comment (snicoll) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentAwareSpringLiquibase.java:55]: You can have two catch blocks here

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentAwareSpringLiquibase.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+				throw (LiquibaseException) ex;
+			}
+			throw new LiquibaseException(ex);
+		}
+	}
+
+	@Override
+	public void destroy() {
+		if (this.environmentId != null) {
+			EnvironmentConfigurationValueProvider.unregisterEnvironment(this.environmentId);
```
Comment (snicoll) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentAwareSpringLiquibase.java:66]: I really would like that the responsability of this was in Liquibase's scope, and not here.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/DataSourceClosingSpringLiquibase.java [hunk 11]
```diff
@@ -58,7 +58,8 @@ private void closeDataSource() {
 	}
 
 	@Override
-	public void destroy() throws Exception {
```
Comment (snicoll) [inline module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/DataSourceClosingSpringLiquibase.java:61]: Please keep the original signature. Even if it's not used by the method's body, it's a breaking change.

**Review by snicoll — CHANGES_REQUESTED** [review]
Thanks for the update. I've added some comments for your consideration.

Also, please run the build as there are checkstyle rules failing atm. For a quick check use `./gradlew :module:spring-boot-liquibase:check`.

File: module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java (hunk 6 above)
Comment (djmiska25) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java:73]: So that’s actually what I did originally, but @wilkinsona raised a concern around how this behaves with child Spring Boot contexts. From what I understand, the way Liquibase handles provider registration makes this tricky. Liquibase’s `ConfigurationValueProvider` registry is global and static, keyed only by class name, while Spring Boot’s configuration model is context-scoped.

Each Spring context can define its own properties and *should* override those from its parent. For example:

**Parent context**
```yaml
spring:
  liquibase:
    properties:
      liquibase.duplicateFileMode: 'WARN'
```

**Child context**
```yaml
spring:
  liquibase:
    properties:
      liquibase.duplicateFileMode: 'ERROR'
```

The expected Boot behavior is that the child context overrides the parent, and the parent keeps its own configuration. But because Liquibase maintains a single global provider list, all providers share the same static state. That’s the core mismatch.

Originally, I just replaced any previously registered provider when a new context started. That works fine for most cases, because Spring Boot typically initializes contexts sequentially and Liquibase itself only runs during startup. This is also the simplest and most test-friendly approach.

The edge case appears when Liquibase is accessed **after startup**-for example, via the Actuator Liquibase endpoint or any custom Liquibase invocation. In those cases, the provider would return values from the most recently initialized Spring context, even if the call originated from the parent context. That’s where correctness can drift.

To get ahead of your specific question: storing the `Environment` directly (as you suggested) works only when there’s a single context. With multiple contexts, there’s no way for Liquibase to indicate *which* Spring context a configuration lookup is coming from. Liquibase doesn’t expose that information. The only feasible mechanism is to pass some form of context identity through Liquibase’s Scope API, which is why the current PR uses that indirection.

And that’s really the broader issue: Liquibase assumes configuration providers are **global and static**, while Boot assumes configuration is **hierarchical and scoped**. Because those models don’t naturally align, we end up with tradeoffs.

So from what I can tell, there are a few realistic paths:

1. **Don’t support overriding in child contexts**, and document that limitation clearly.  
2. **Partially support it** (current PR behavior): initialization + the Actuator endpoint work correctly, but any ad-hoc Liquibase usage might not resolve the right context. Practically speaking, those scenarios seem very rare.
3. **Decide the complexity isn’t worth it and close the PR.**

I’m open to whichever direction best fits Boot’s conventions. My main goal here is just to make the tradeoffs visible so we can decide on the right level of support.

Comment (snicoll) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java:73]: > So that’s actually what I did originally, but @wilkinsona raised a concern around how this behaves with child Spring Boot contexts.

Maybe I missed something but I was suggesting to put the `Environment` keyed by its `ApplicationContext#getId` in the `Scope`. This would avoid the indirection and the static cache. WDYT?

If that's a sensible option, you could prefix the key to make it more explicit, like `application-id-<theid>`.

Comment (djmiska25) [reply module/spring-boot-liquibase/src/main/java/org/springframework/boot/liquibase/autoconfigure/EnvironmentConfigurationValueProvider.java:73]: Oh ok, I think I understand now. So we could, if your preference is not to have any fallback options, to infer the environment. Like I mentioned above, there are some edge cases that could occur **outside** of the liquibase scope. If that happens currently, we just return null and skip this provider. However, its possible to still infer the value in some scenarios, such as:
- There's only one env in the local cache
- The value is the same among all the env's in the cache

The only real time that we couldn't fully handle it is if the value differs among multiple env's and we're outside a liquibase scope. If we store the env fully in the liquibase scope we can't do any inference. But, I do think this could be more difficult to document or maintain so i'm curious on your opinion on this? I personally, don't really know how often people do custom liquibase instances, so I'm not sure how important it might be.

But, we could definitely store it in the scope; there would just 100% be no way to fall back because we'd have no access to the env at all outside the Liquibase scope.

**Comment by wilkinsona** [comment]
Thank you for all our efforts here, @dmiska25. They're really helped to flesh out the pros and cons of various approaches. On balance, I'm unfortunately now of the opinion that we should close this and some changes to `SpringLiquibase` should be pursued instead. 

`SpringLiquibase` already provides a "Spring-ified wrapper for Liquibase". What it's missing is integration with Spring's `Environment`. If it was `EnvironmentAware`, it could offer a mechanism to perform operations in such as way that the `Environment` can provide configuration values while the operation is running. This mechanism could be used both in `afterPropertiesSet()` for the DB migration doing startup as well as by things like Boot's Liquibase endpoint. This would also mean that any application code that wants to make ad-hoc calls to Liquibase work as expected in terms of `Environment`-provided configuration as long as they went through `SpringLiquibase` and the new mechanism.

@dmiska25, would you like to comment on https://github.com/liquibase/liquibase/issues/3230 and see if it helps to move that issue forward? I've subscribed to the issue and will do what I can to help.
